### PR TITLE
aws backend skeleton

### PR DESCRIPTION
**Pull request by beining00**
### Skeleton code for aws backend. 
* template.yml - provision all was resources includes, 
    * two lambda functions 
    * dynamodb table
    * S3 bucket 
    * API endpoint 
    * SQS trigger for postProcessHandler
 
* two lamdbas 
    * postProcessHandler
    * uploadHandler

### Test set up 
* test files has the name convention `*.test.ts`
* `jest` framework is used. 

### Usage 
1. install `aws cli` and `aws sam`
2. configure your local AWS profile 
2. sam build (in the same directory as template.yml) 
3. sam deploy

### CODE REVIEW

File: server/uploadHandler/package.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "web_template",
+  "version": "1.0.0",
+  "author": "SAM CLI",
+  "license": "MIT",
```
Comment (idomo1) [inline server/uploadHandler/package.json:5]: Please update the name, author and license

File: server/postProcessHandler/package.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "web_template",
+  "version": "1.0.0",
+  "author": "SAM CLI",
+  "license": "MIT",
```
Comment (idomo1) [inline server/postProcessHandler/package.json:5]: Please update the name, author and license

File: server/uploadHandler/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        console.log('Lambda environment variables is missing the SAMPLE_TABLE variable required.');
+        return { statusCode: 500 };
+    }
+
+    console.log("job_table",process.env['JOB_TABLE'])
+    console.log("src_bucket",process.env['SRC_BUCKET'])
+    console.log("region",process.env['AWS_REGION'])
+
+    await createNewJob()
+    console.log('Running uploadHandler');
```
Comment (idomo1) [inline server/uploadHandler/src/index.ts:24]: The uploadHandler is going to be called inside of `createNewJob` right? If so, then the log should be inside that function (or just removed until it is actually implemented)

File: server/uploadHandler/src/service/dynamodbClient.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+export const putItem = (
+  params: PutItemInput
+): Promise<PromiseResult<PutItemOutput, AWSError>> =>
+  dynamoDb.putItem(params).promise()
+
+export const updateItem = (
+  params: UpdateItemInput
+): Promise<PromiseResult<UpdateItemOutput, AWSError>> =>
+  dynamoDb.updateItem(params).promise()
```
Comment (idomo1) [inline server/uploadHandler/src/service/dynamodbClient.ts:44]: Is there a reason to have this duplicated for the uploadHandler and postProcessHandler? Can't we have one file which both services import?

**Review by idomo1 — CHANGES_REQUESTED** [review]

**Review by YinghaoMa — APPROVED** [review]

File: server/uploadHandler/src/service/dynamodbClient.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+} from "aws-sdk/clients/dynamodb";
+import { PromiseResult } from "aws-sdk/lib/request";
+import { Agent } from "https";
+
+const awsRegion = process.env.AWS_REGION
+
+// The reason of using DynamoDB directly, see:
+// https://github.com/aws/aws-sdk-js/issues/3578
+const dynamoDb = new DynamoDB({
+  apiVersion: "2012-08-10",
```
Comment (reteppeter) [inline server/uploadHandler/src/service/dynamodbClient.ts:20]: Same issue as other use.

File: server/uploadHandler/src/service/dynamodbClient.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  UpdateItemInput,
+  UpdateItemOutput,
+} from "aws-sdk/clients/dynamodb";
+import { PromiseResult } from "aws-sdk/lib/request";
+import { Agent } from "https";
+
+const awsRegion = process.env.AWS_REGION
+
+// The reason of using DynamoDB directly, see:
+// https://github.com/aws/aws-sdk-js/issues/3578
```
Comment (reteppeter) [inline server/uploadHandler/src/service/dynamodbClient.ts:18]: I'm unsure how that GitHub issue relates to this code.

File: server/uploadHandler/src/service/dynamodbService.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    imageFiles: string[],
+  ): Promise<void> => {
+    const newJobItem = {
+      TableName: tableName,
+      Item:{
+          id: uuidv4,
+          email,
+          videoFiles,
+          imageFiles,
+          updatedAt: new Date().toISOString()
```
Comment (reteppeter) [inline server/uploadHandler/src/service/dynamodbService.ts:20]: Job start and end times would probably be good to have as well.

File: server/postProcessHandler/src/service/dynamodbService.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+//           id: uuidv4,
+//           email,
+//           videoFiles,
+//           imageFiles,
+//           updatedAt: new Date().toISOString()
+//       } 
+//     } as PutItemInput
+  
+//     await putItem(newJobItem)
+//   }
```
Comment (reteppeter) [inline server/postProcessHandler/src/service/dynamodbService.ts:24]: Except for the imports the commented out code in this file is just a copy of server/uploadHandler/src/service/dynamodbService.ts, so should be removed.

**Review by reteppeter — CHANGES_REQUESTED** [review]

File: server/postProcessHandler/src/service/dynamodbClient.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+} from "aws-sdk/clients/dynamodb";
+import { PromiseResult } from "aws-sdk/lib/request";
+import { Agent } from "https";
+
+const awsRegion = process.env.AWS_REGION
+
+// The reason of using DynamoDB directly, see:
+// https://github.com/aws/aws-sdk-js/issues/3578
+const dynamoDb = new DynamoDB({
+  apiVersion: "2012-08-10",
```
[the remark this answers is not in the document]
Comment (idomo1) [reply server/postProcessHandler/src/service/dynamodbClient.ts:20]: From the AWS documentation:
```
```

File: server/uploadHandler/src/service/dynamodbClient.ts (hunk 4 above)
Comment (beining00) [reply server/uploadHandler/src/service/dynamodbClient.ts:44]: Emmm.... my understand of lamdba is the each lambda function is completely separate from each other. I provisioned the `uploadHandler` and the `PostProcessHandler` as two different lamdba, so I guess I need to have a copy of it in both folder.

File: server/uploadHandler/src/createNewJob.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { addNewJobToDb } from "./service/dynamodbService";
+
+export const createNewJob = async () => {
+
+    // 1. upload files to S3
```
[the remark this answers is not in the document]
Comment (beining00) [reply server/uploadHandler/src/createNewJob.ts:5]: For now, we might just delete old files manually in the S3 bucket.

File: server/postProcessHandler/src/service/dynamodbService.ts (hunk 8 above)
Comment (beining00) [reply server/postProcessHandler/src/service/dynamodbService.ts:24]: Yeah I should delete the commented out code. The reason why I can't import it from `uploadHandler` is explained [here](https://github.com/Visual-Testing-for-Android-Apps/backend/pull/12#discussion_r680648910).

File: server/uploadHandler/package.json (hunk 1 above)
Comment (beining00) [reply server/uploadHandler/package.json:5]: what should I change this to ?

File: server/uploadHandler/src/service/dynamodbClient.ts (hunk 6 above)
Comment (beining00) [reply server/uploadHandler/src/service/dynamodbClient.ts:18]: Commets Removed.

File: server/uploadHandler/package.json (hunk 1 above)
Comment (idomo1) [reply server/uploadHandler/package.json:5]: Remove the license or change it to an empty string. Maybe change the name to "upload_hander" and change the author to "VTAA"?

File: server/postProcessHandler/src/service/dynamodbClient.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (reteppeter) [reply server/postProcessHandler/src/service/dynamodbClient.ts:19]: As long as 2012 is the most recent version I'm fine with this then.

File: server/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
+# sam-app
+
+This template generates a sample NodeJS 12.x project configured with TypeScript using the [Serverless Application Model (SAM)](https://github.com/awslabs/serverless-application-model).
```
Comment (mt-empty) [inline server/README.md:3]: The template uses `NodeJS 12`, although it build with no errors, @idomo1 will this be a problem in the future ?

**Comment by mt-empty** [comment]
So this commit https://github.com/Visual-Testing-for-Android-Apps/backend/pull/12/commits/75b8e748223d20efd75b4386f3e8d6ae3ba8ddf0 is all skeleton right? that means I don't have to focus on it?

File: server/README.md (hunk 11 above)
Comment (idomo1) [reply server/README.md:3]: CI might break at some point, we can deal with that when/if it happens

File: server/uploadHandler/src/service/dynamodbService.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  imageFiles: string[]
+): Promise<void> => {
+  const newJobItem = {
+    TableName: tableName,
+    Item: {
+      id: uuidv4,
+      email,
+      videoFiles,
+      imageFiles,
+      updatedAt: new Date().toISOString(),
```
Comment (idomo1) [inline server/uploadHandler/src/service/dynamodbService.ts:20]: @beining00 Is the aim of this PR to finalise the database schema? Or will that be done later?
One thing missing from this is the status of each image/video

File: server/README.md (hunk 11 above)
Comment (mt-empty) [reply server/README.md:3]: Not really concerned about CI/CD, I was more concerned about compatibility issues

Comment (idomo1) [reply server/README.md:3]: I don't know the particulars of the differences between Node 12 and 14/15, but I'm not that concerned we'll run into issues

**Review by idomo1 — APPROVED** [review]

**Review by reteppeter — APPROVED** [review]

File: server/uploadHandler/package.json [hunk 13]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "vtaa",
```
Comment (mt-empty) [inline server/uploadHandler/package.json:2]: Should the upload handler be name `vtaa`. Also its worth noting that the root `package.json` is named `visiondroid`

**Review by mt-empty — CHANGES_REQUESTED** [review]

**Review by mt-empty — APPROVED** [review]

**Review by reteppeter — APPROVED** [review]

File: server/uploadHandler/src/service/dynamodbService.ts (hunk 12 above)
Comment (beining00) [reply server/uploadHandler/src/service/dynamodbService.ts:20]: I wrote this just to give an example of how to update the database. The actual fields in the table can be changed later on.

**Pull request merged**